rust/foxglove/src/websocket.rs

```
@@ -821,13 +892,31 @@ impl Server {
                     if self.started.load(Acquire) {
                         tracing::error!("Error sending data message to client {addr}: {err}");
                     } else {
-                        new_client.data_plane_rx.drain();
-                        new_client.control_plane_rx.drain();
+                        new_client.drain_queues();
                     }
                 }
             }
         };
 
+        let send_service_call_responses = async {
+            if let Some(chan) = new_client.service_response_chan.as_ref() {
+                loop {
+                    let msg = chan.next_message().await;
+                    let mut sender = new_client.sender.lock().await;
+                    if let Err(err) = sender.send(msg).await {
+                        if self.started.load(Acquire) {
+                            tracing::error!("Error sending data message to client {addr}: {err}");
+                        } else {
+                            new_client.drain_queues();
+                        }
+                    }
+                }
+            } else {
+                // No services capability, no problem.
+                std::future::pending().await
+            };
+        };
+
```

Can't we use the control plane for service messages via send_control_msg? I don't see why we need a third type of queue here. The behavior of try_send or disconnect is fine for service calls too.

---

I had a few things that I wanted:

- Limit the absolute service call concurrency on a per-client basis.
- Fail fast, instead of doing a bunch of throw-away work in the service call handler.
- Give clients have an opportunity to recover from "too many requests", instead of just disconnecting them. This is a different situation from a client that can't keep up with the rate of messages it's receiving from the server.

I also figured that we'll eventually rework some of our outbound queuing, so I didn't want to make changes to the existing queues. Maybe we can revisit this in the future, and find a way to merge some of these concepts?

---

Also, I think we want to try _VERY_ hard to deliver service call responses.

Consider the implementation of the client. When the client makes a call, they're awaiting a response. If the server silently drops the response because its queues happen to be full, the client will block forever.

---

I think you still achieve all of those objectives - I'm not asking to change the ticket system for the request semaphore.

The difference is that if the control queue fills, regardless of what the source of those messages are, we disconnect the client. The message doesn't get dropped like with the data plane. I think this is fine, because the service call limit is so much lower than the control plane queue size that it would only happen for slow clients that are backed up.

The disconnect instead of drop is not implemented yet, but should next week with FG-10441. And we should increase the control queue size to a much larger size in the same PR (and maybe give a way to configure it in the server builder.)

---

OK, that's fair. I'll drop the `ResponseChannel` and replace it with a simple non-blocking semaphore for admission control.

---

I've implemented this. But to be honest, I still mildly prefer the old solution, with the dedicated service call response channel. It's more loosely-coupled and modular, since the client doesn't need to care about the semaphore, and the responder doesn't need to care about the client. While it is "one more thing to poll" in the client, I think service calls are sufficiently different from other styles of client interaction to merit their own queue.